<a href="https://colab.research.google.com/github/ipshitaRB/Analyze-Employee-Exit-Surveys/blob/master/Analyze_Employee_Exit_Surveys.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clean and Analyze Employee Exit Surveys 

In this project, we will work with exit surveys from employees of the  [Department of Education, Training and Employment (DETE)](https://en.wikipedia.org/wiki/Department_of_Education_and_Training_(Queensland)) and the Technical and Further Education (TAFE) institute in Queensland, Australia. You can find the TAFE exit survey [here](https://data.gov.au/dataset/ds-qld-89970a3b-182b-41ea-aea2-6f9f17https://data.gov.au/dataset/ds-qld-fe96ff30-d157-4a81-851d-215f2a0fe26d/details?q=exit%20survey).

Below is a preview of a couple columns we'll work with from the dete_survey.csv:

* ID: An id used to identify the participant of the survey
* SeparationType: The reason why the person's employment ended
* Cease Date: The year or month the person's employment ended
* DETE Start Date: The year the person began employment with the DETE

Below is a preview of a couple columns we'll work with from the tafe_survey.csv:

* Record ID: An id used to identify the participant of the survey
* Reason for ceasing employment: The reason why the person's employment ended
* LengthofServiceOverall. Overall Length of Service at Institute (in years): The length of the person's employment (in years)

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [0]:
import pandas as pd
import numpy as np


## Load data into dataframes

In [0]:
dete_survey = pd.read_csv('/content/drive/My Drive/Data Science /Data/Employee Exit Surveys/dete_survey.csv')
tafe_survey = pd.read_csv('/content/drive/My Drive/Data Science /Data/Employee Exit Surveys/tafe_survey.csv')

## Explore dete_survey and tafe_survey

### Explore dete_survey

In [4]:
dete_survey.head()

ID                    SeparationType Cease Date  ... South Sea Disability NESB
0   1             Ill Health Retirement    08/2012  ...       NaN        NaN  Yes
1   2  Voluntary Early Retirement (VER)    08/2012  ...       NaN        NaN  NaN
2   3  Voluntary Early Retirement (VER)    05/2012  ...       NaN        NaN  NaN
3   4         Resignation-Other reasons    05/2012  ...       NaN        NaN  NaN
4   5                    Age Retirement    05/2012  ...       NaN        NaN  NaN

[5 rows x 56 columns]

Some columns such as 'Role Start Date' and 'DETE Start Date' have values such as 'Not Stated' instead of NaN. 

* A - Agree
* N - Neutral
* D - Disagree
* SD - Strongly Disagree
* SA - Strongly Agree

In [5]:
dete_survey.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
ID                                     822 non-null int64
SeparationType                         822 non-null object
Cease Date                             822 non-null object
DETE Start Date                        822 non-null object
Role Start Date                        822 non-null object
Position                               817 non-null object
Classification                         455 non-null object
Region                                 822 non-null object
Business Unit                          126 non-null object
Employment Status                      817 non-null object
Career move to public sector           822 non-null bool
Career move to private sector          822 non-null bool
Interpersonal conflicts                822 non-null bool
Job dissatisfaction                    822 non-null bool
Dissatisfaction with the department    822 non-null bool
Physical work environ

In [6]:
dete_survey['SeparationType'].value_counts()

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: SeparationType, dtype: int64

In [7]:
dete_survey.isnull().sum()

ID                                       0
SeparationType                           0
Cease Date                               0
DETE Start Date                          0
Role Start Date                          0
Position                                 5
Classification                         367
Region                                   0
Business Unit                          696
Employment Status                        5
Career move to public sector             0
Career move to private sector            0
Interpersonal conflicts                  0
Job dissatisfaction                      0
Dissatisfaction with the department      0
Physical work environment                0
Lack of recognition                      0
Lack of job security                     0
Work location                            0
Employment conditions                    0
Maternity/family                         0
Relocation                               0
Study/Travel                             0
Ill Health 

Some columns such as Torres Strait and South Sea have more than 400 non-null values out of 822. We should explore how replacing the null values and dropping such columns would impact the analysis.

### Explore tafe_survey


In [8]:
tafe_survey.head()

Record ID  ... LengthofServiceCurrent. Length of Service at current workplace (in years)
0  6.341330e+17  ...                                                1-2                       
1  6.341337e+17  ...                                                NaN                       
2  6.341388e+17  ...                                                NaN                       
3  6.341399e+17  ...                                                NaN                       
4  6.341466e+17  ...                                                3-4                       

[5 rows x 72 columns]

In [9]:
tafe_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
Record ID                                                                                                                                                        702 non-null float64
Institute                                                                                                                                                        702 non-null object
WorkArea                                                                                                                                                         702 non-null object
CESSATION YEAR                                                                                                                                                   695 non-null float64
Reason for ceasing employment                                                                                                                                    701 non-

The column names need to be renamed with more concise names. The columns can be transformed into variables with their corresponding values.

In [10]:
tafe_survey['Reason for ceasing employment'].value_counts()

Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: Reason for ceasing employment, dtype: int64

In [11]:
tafe_survey.isnull().sum()

Record ID                                                                      0
Institute                                                                      0
WorkArea                                                                       0
CESSATION YEAR                                                                 7
Reason for ceasing employment                                                  1
                                                                            ... 
CurrentAge. Current Age                                                      106
Employment Type. Employment Type                                             106
Classification. Classification                                               106
LengthofServiceOverall. Overall Length of Service at Institute (in years)    106
LengthofServiceCurrent. Length of Service at current workplace (in years)    106
Length: 72, dtype: int64

## Identify Missing Values 

We will find "Not Stated" values and represent them as NaN

In [13]:
dete_survey = pd.read_csv('/content/drive/My Drive/Data Science /Data/Employee Exit Surveys/dete_survey.csv', na_values= 'Not Stated')
dete_survey.head()

ID                    SeparationType Cease Date  ...  South Sea  Disability NESB
0   1             Ill Health Retirement    08/2012  ...        NaN         NaN  Yes
1   2  Voluntary Early Retirement (VER)    08/2012  ...        NaN         NaN  NaN
2   3  Voluntary Early Retirement (VER)    05/2012  ...        NaN         NaN  NaN
3   4         Resignation-Other reasons    05/2012  ...        NaN         NaN  NaN
4   5                    Age Retirement    05/2012  ...        NaN         NaN  NaN

[5 rows x 56 columns]

## Drop Unnecessary Columns
Drop columns that are not needed for analysis

In [14]:
# Remove columns we don't need for our analysis
dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49], axis=1)
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66], axis=1)

#Check that the columns were dropped
print(dete_survey_updated.columns)
print(tafe_survey_updated.columns)

Index(['ID', 'SeparationType', 'Cease Date', 'DETE Start Date',
       'Role Start Date', 'Position', 'Classification', 'Region',
       'Business Unit', 'Employment Status', 'Career move to public sector',
       'Career move to private sector', 'Interpersonal conflicts',
       'Job dissatisfaction', 'Dissatisfaction with the department',
       'Physical work environment', 'Lack of recognition',
       'Lack of job security', 'Work location', 'Employment conditions',
       'Maternity/family', 'Relocation', 'Study/Travel', 'Ill Health',
       'Traumatic incident', 'Work life balance', 'Workload',
       'None of the above', 'Gender', 'Age', 'Aboriginal', 'Torres Strait',
       'South Sea', 'Disability', 'NESB'],
      dtype='object')
Index(['Record ID', 'Institute', 'WorkArea', 'CESSATION YEAR',
       'Reason for ceasing employment',
       'Contributing Factors. Career Move - Public Sector ',
       'Contributing Factors. Career Move - Private Sector ',
       'Contributing Fact

## Update Column names
Each dataframe contains many of the same columns, but the column names are different. Below are some of the columns we'd like to use for our final analysis:



|dete_survey|tafe_survey|Definition|
|-----------|------------|----------|
|ID	|Record ID	|An id used to identify the participant of the survey|
SeparationType|	Reason for ceasing employment|The reason why the participant's employment ended
Cease Date|	CESSATION YEAR|	The year or month the participant's employment ended
DETE Start Date	| |The year the participant began employment with the DETE
||LengthofServiceOverall.Overall Length of Service at Institute (in years)|The length of the person's employment (in years)
Age	|CurrentAge.Current Age|The age of the participant
Gender|	Gender.What is your Gender?|The gender of the participant


Because we eventually want to combine them, we'll have to standardize the column names.

In [15]:
dete_survey_updated.columns = dete_survey_updated.columns.str.strip().str.replace(" ","_").str.lower()
print(dete_survey_updated.columns)

Index(['id', 'separationtype', 'cease_date', 'dete_start_date',
       'role_start_date', 'position', 'classification', 'region',
       'business_unit', 'employment_status', 'career_move_to_public_sector',
       'career_move_to_private_sector', 'interpersonal_conflicts',
       'job_dissatisfaction', 'dissatisfaction_with_the_department',
       'physical_work_environment', 'lack_of_recognition',
       'lack_of_job_security', 'work_location', 'employment_conditions',
       'maternity/family', 'relocation', 'study/travel', 'ill_health',
       'traumatic_incident', 'work_life_balance', 'workload',
       'none_of_the_above', 'gender', 'age', 'aboriginal', 'torres_strait',
       'south_sea', 'disability', 'nesb'],
      dtype='object')


### Rename TAFE column names
We are renaming TAFE column names to match DETE column names before they are combined.


In [0]:
new_column_names = {'Record ID': 'id',
'CESSATION YEAR': 'cease_date',
'Reason for ceasing employment': 'separationtype',
'Gender. What is your Gender?': 'gender',
'CurrentAge. Current Age': 'age',
'Employment Type. Employment Type': 'employment_status',
'Classification. Classification': 'position',
'LengthofServiceOverall. Overall Length of Service at Institute (in years)': 'institute_service',
'LengthofServiceCurrent. Length of Service at current workplace (in years)': 'role_service'}

tafe_survey_updated = tafe_survey_updated.rename(columns=new_column_names)


In [18]:
# check that column names are renamed
print(tafe_survey_updated.columns)

Index(['id', 'Institute', 'WorkArea', 'cease_date', 'separationtype',
       'Contributing Factors. Career Move - Public Sector ',
       'Contributing Factors. Career Move - Private Sector ',
       'Contributing Factors. Career Move - Self-employment',
       'Contributing Factors. Ill Health',
       'Contributing Factors. Maternity/Family',
       'Contributing Factors. Dissatisfaction',
       'Contributing Factors. Job Dissatisfaction',
       'Contributing Factors. Interpersonal Conflict',
       'Contributing Factors. Study', 'Contributing Factors. Travel',
       'Contributing Factors. Other', 'Contributing Factors. NONE', 'gender',
       'age', 'employment_status', 'position', 'institute_service',
       'role_service'],
      dtype='object')
